exporter/fileexporter/file_exporter.go

```diff
@@ -70,55 +69,112 @@ func (e *fileExporter) consumeTraces(_ context.Context, td ptrace.Traces) error
 	if err != nil {
 		return err
 	}
-	buf = e.compressor(buf)
-	return e.exporter(e, buf)
+	_, err = e.exporter.Write(buf)
+	return err
 }
 
 func (e *fileExporter) consumeMetrics(_ context.Context, md pmetric.Metrics) error {
 	buf, err := e.metricsMarshaler.MarshalMetrics(md)
 	if err != nil {
 		return err
 	}
-	buf = e.compressor(buf)
-	return e.exporter(e, buf)
+	_, err = e.exporter.Write(buf)
+	return err
 }
 
 func (e *fileExporter) consumeLogs(_ context.Context, ld plog.Logs) error {
 	buf, err := e.logsMarshaler.MarshalLogs(ld)
 	if err != nil {
 		return err
 	}
-	buf = e.compressor(buf)
-	return e.exporter(e, buf)
+	_, err = e.exporter.Write(buf)
+	return err
+}
+
+type lineWriter struct {
+	mutex sync.Mutex
+	file  io.Writer
+}
+
+func NewLineWriter(cfg *Config, file io.Writer) io.Writer {
+	if cfg.Compression == "zstd" {
+		if fw, err := zstd.NewWriter(file); err == nil {
+			// flushing the compressed writer every second.
+			go func() {
+				for {
+					time.Sleep(1 * time.Second)
+					if fw.Flush() != nil {
+						return
+					}
+				}
+			}()
+
+			return &lineWriter{
+				file: bufio.NewWriter(fw),
+			}
+		}
+	}
+
+	return &lineWriter{
+		file: bufio.NewWriter(file),
+	}
 }
 
-func exportMessageAsLine(e *fileExporter, buf []byte) error {
+func (lw *lineWriter) Write(buf []byte) (int, error) {
 	// Ensure only one write operation happens at a time.
-	e.mutex.Lock()
-	defer e.mutex.Unlock()
-	if _, err := e.file.Write(buf); err != nil {
-		return err
+	lw.mutex.Lock()
+	defer lw.mutex.Unlock()
+	if _, err := lw.file.Write(buf); err != nil {
+		return 0, err
 	}
-	if _, err := io.WriteString(e.file, "\n"); err != nil {
-		return err
+	if _, err := io.WriteString(lw.file, "\n"); err != nil {
+		return 0, err
 	}
-	return nil
+	return 1 + len(buf), nil
 }
 
-func exportMessageAsBuffer(e *fileExporter, buf []byte) error {
+type fileWriter struct {
+	mutex sync.Mutex
+	file  io.Writer
+}
+
+func NewFileWriter(cfg *Config, file io.Writer) io.Writer {
+	if cfg.Compression == "zstd" {
+		if fw, err := zstd.NewWriter(file); err == nil {
+			// flushing the compressed writer every second.
+			go func() {
+				for {
+					time.Sleep(1 * time.Second)
+					if fw.Flush() != nil {
+						return
+					}
+				}
+			}()
+			return &fileWriter{
+				file: bufio.NewWriter(fw),
+			}
+		}
+	}
```

Comment:
This changes compression to be per-stream not per-record, which I think is OK as long as the reader (a) has never supported this, (b) will support this after your change.